pkg/mutation/path/path.go

```diff
@@ -0,0 +1,32 @@
+package path
+
+// Parse parses the string representing a path specification
+// and returns the path splitted as path entries
+func Parse(string) ([]Entry, error) {
+	return nil, nil
+}
+
+//Type is the type of a path
+type Type string
+
+// Entry represent a segment of a path
+type Entry interface {
+	Type() Type
+}
+
+// Object is a segment of a path which is not a list
+type Object struct {
+	// TypeName is the type of Entry
+	Type Type
+	// The name of the field corresponding to the next segment of the path.
+	// Must be an empty string in case
+	PointsTo *string
```

Comment:
What is this for? Are we trying to construct a linked list of the path entries? Why is that needed?

Reply:
Isn't this needed for traversing the object?
PointsTo is the name of the field of the object the next item refers to.

Reply:
I imagined we would be passing `[]path.Entry` around, but I haven't looked at all the in-flight PRs yet.

Reply:
Yep, and each element of `[]path.Entry` identifies a segment (or a list) of the path. PointsTo is the name of the field related to that given segment (or, at least is how I understood it)

Reply:
Ok I understand - yes we need a Name/Text/Value/Whatever field for the segment. I think this also doesn't need to be a pointer. The comment referring to the _next segment_ was throwing me off - can we try a concrete example?

Input: `.spec.containers[name:*].securityContext`
Expected:
```
[]Entry{
  Object{
    Type: "Object",
    Value: "spec",
  },
  Object{
    Type: "Object",
    Value: "containers",
  },
  List{
    Type: "List",
    KeyField: "name",
    KeyValue: "*",
    Globbed: true,
  },
  Object{
    Type: "Object",
    Value: "securityContext",
  },
}
```

-^ does this look right? Each node describes itself and not the next node. Also I've omitted a root node representation, does this sound reasonable?

Reply:
It does.
Re: pointer vs string, there was this discussion between you and @maxsmythe : https://docs.google.com/document/d/1c5Z3g6Zsfmga7xod4--t6RrXamdAaMaqYzPNujtZb10/edit?disco=AAAAHLfZaLE

For the key, the idea was to enforce the field not being there in case of globs.
For the path, max's comment was to distinguish between an empty string path and an unset one, but now I can't see how we could have unset paths in the middle if []Entry

Reply:
re: how to get unset paths, I don't think we should ever have unset paths. The expected parsing in the case above would be:

Input: .spec.containers[name:*].securityContext
Expected:
```
[]Entry{
  Object{
    Type: "Object",
    PointsTo: "spec",
  },
  Object{
    Type: "Object",
    PointsTo: "containers",
  },
  List{
    Type: "List",
    KeyField: "name",
    KeyValue: "", (or nil if this is a pointer) -- we need to disambiguate between "*" and *
    Globbed: true,
  },
  Object{
    Type: "Object",
    PointsTo: "securityContext",
  },
}
```

Reply:
Sorry, posted before complete.

The two big differences between my example and Oren's are `Value` -> `PointsTo` and the treatment of globs

Oren, you actually did include a root node in your example. Your root node points to "spec".

The idea here is that each entry in the list provides type information for the current position we are at. In the above example:

* 0th node is an object that has a field named "spec"
* 1st node is an object (called "spec" per the previous node) that has a field called "containers"
* 2nd node, "containers", is a list of objects with keyfield "name"
* 3rd node is an object that has a field named "securityContext", we can also infer the string field "name"

The reason for "PointsTo" is that we expect the last item in the `location` to be ambiguously typed -- it will take on whatever value the user supplies in their mutation.

Reply:
Thanks for the breakdown, Max. What would the ambiguously typed last field of a location look like in string form?

Reply:
It just wouldn't show up, take `spec.containers[name: *].securityContext` for example:

The user could provide the value `"Gee I hope I can use a string here"`, `7` or `{"make": "secure"}`

for `securityContext` and the location would shed no light on which one of these are correct.

Reply:
> 	// Must be an empty string in case of the last item of a []Entry
	PointsTo *string

Using the `spec.containers[name: *].securityContext` example from above, are we missing the last item after: 
```
 Object{
    Type: "Object",
    PointsTo: "securityContext",
  },
```

Or does this comment need to be updated?

Reply:
I _think_ we can update the comment. I took it directly from Max's doc, but I am not sure adding a

 Object{
    Type: "Object",
    PointsTo: "",
  },

 empty object would add value here, especially looking at the examples.

Reply:
@maxsmythe I removed the comment as per @ritazh 's suggestion, just let me know if I missed something.

Reply:
For Object nodes, I am now pretty sure we don't need the optional-points-to-ness. See the following example:

`spec."".foo`: this looks up the empty-string field name in `spec`.
`spec..foo`: this is a parsing error, the whole path will be nil.

Reply:
Couple things:

1) Yeah, that "last entry" comment was definitely wrong, it's good we removed it.

2) I'm fine with removing the pointers for now.

Reply:
Removed the pointer from `Object`